### PR TITLE
Feature: dxDrawModel3d

### PR DESCRIPTION
**Pull request by CrosRoad95**
This pull request adds function to draw arbitrary model in simillar fasion as other dxDraw* functions

syntax: `bool dxDrawModel3d(model, position, rotation, optional scale)` ( draw in onClientRender/onClientPreRender loop )

You can draw as many objects as you want, it has no limits both amount of models and draw distance.
Rendered models has no collision, they look exactly the same as objects created using "createObject" ( possibly transparency will be an issue )
use cases:
1. draw vegetation
2. draw ped/vehicles visual accessories
3. draw miscellaneous objects on ground as an items player can pick up

example: `dxDrawModel3d(3276, -719.64984, 951.26685, 12.13281, 0, 0,90)` - as you can see, it look 1:1 as createObject
image
image

No limits:
image
image

### CODE REVIEW

**Comment by Einheit-101** [comment]
This is nice. I guess you trade performance for free object streaming slots?

**Comment by CrosRoad95** [comment]
you can call it that

**Comment by Pirulax** [comment]
Eh I'd argue this is faster than having objects, as they have a lot of overhead.
These don't have collisions either, so there's that.

File: Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp [hunk 1]
```diff
@@ -64,6 +75,7 @@ void CLuaDrawingDefs::LoadFunctions()
         {"dxSetAspectRatioAdjustmentEnabled", DxSetAspectRatioAdjustmentEnabled},
         {"dxIsAspectRatioAdjustmentEnabled", DxIsAspectRatioAdjustmentEnabled},
         {"dxSetTextureEdge", DxSetTextureEdge},
+        {"dxDrawModel3d", ArgumentParser<DxDrawModel3d>},
```
Comment (Lpsd) [inline Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp]: rename to `dxDrawModel3D` for case consistency with other functions

**Review by Lpsd — CHANGES_REQUESTED** [review]

**Review by Lpsd — APPROVED** [review]

**Comment by tederis** [comment]
There is a lot of commented out lines. I doubt that it should be merged in this "dirty" condition. But now it is too late to worry anyway.

**Comment by Lpsd** [comment]
It's never too late to worry. https://github.com/multitheftauto/mtasa-blue/pull/3210

**Comment by onikyannn** [comment]
Object is rendered at zero coordinates in any dimension other than 0.

**Comment by Xenius97** [comment]
What about dxDrawModel2D?
Non shadered version of https://wiki.multitheftauto.com/wiki/Resource:Object_preview

**Comment by tederis** [comment]
> What about dxDrawModel2D? Non shadered version of https://wiki.multitheftauto.com/wiki/Resource:Object_preview

It will be enough just to call dxDrawModel3d inside a render target.

**Comment by PlatinMTA** [comment]
> It will be enough just to call dxDrawModel3d inside a render target.

wiki says no rendertarget support

**Comment by tederis** [comment]
> > It will be enough just to call dxDrawModel3d inside a render target.
> 
> wiki says no rendertarget support

Of course there is no RT support now. But it can be easily added.

**Comment by Dutchman101** [comment]
Reverted in 5b44a09 as a result of dev discord feedback:

_TEDERIs — Today at 2:24 PM
I respect the efforts of CrosRoad95. But I think #3172 (dxDrawModel3d feature) should be revoked. It's very buggy now, has a lot of internal problems and even cause crashes. It renders an object randomly (due to the nature of CClientGame::GameEntityRenderHandler) which leads to an unpredictable appearience(and can even disappear). Even after #3210 it contains unused code(see HOOKPOS_CRenderer__RenderFadingInEntities_MID). And it can crash the game(see model handling in DxDrawModel3D). This is nice PR, but it still has a long way to go before it is ready for production use_

**Pull request merged**